File: server/routes/vehicles.py
"""
Traffic Violation Management System — Vehicle Search Routes
Self-contained with pymysql, for police use only
"""
from fastapi import APIRouter, HTTPException, status
import pymysql

router = APIRouter()

# Database configuration
DB_CONFIG = {
    'host': '127.0.0.1',
    'user': 'root',
    'password': 'yvpandi@11',
    'database': 'traffic_violation_db',
    'port': 3306,
    'connect_timeout': 5,
    'read_timeout': 10,
    'write_timeout': 10,
    'cursorclass': pymysql.cursors.DictCursor
}


def get_db_connection():
    """Get database connection."""
    try:
        conn = pymysql.connect(**DB_CONFIG)
        return conn
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Database connection failed: {str(e)}"
        )


@router.get("/search/{plate_no}")
async def search_vehicle(plate_no: str):
    """Search vehicle by plate number and get full violation history."""
    conn = None
    cursor = None
    
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        
        # Get vehicle details
        cursor.execute(
            """SELECT 
                   plate_no,
                   vehicle_model,
                   vehicle_type,
                   owner_name,
                   owner_type,
                   registered_at
               FROM VEHICLES
               WHERE plate_no = %s""",
            (plate_no,)
        )
        vehicle = cursor.fetchone()
        
        if not vehicle:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Vehicle with plate number '{plate_no}' not found"
            )
        
        # Convert datetime to string
        if vehicle.get('registered_at'):
            vehicle['registered_at'] = vehicle['registered_at'].isoformat()
        
        # Get violation history with challans
        cursor.execute(
            """SELECT 
                   ve.event_id,
                   ve.event_timestamp,
                   ve.location_coords,
                   ve.notes,
                   vr.rule_code,
                   vr.rule_name,
                   vr.base_fine_amount,
                   vr.severity,
                   c.challan_id,
                   c.total_amount,
                   c.payment_status,
                   c.issue_date,
                   c.due_date,
                   c.paid_at
               FROM VIOLATION_EVENTS ve
               JOIN VIOLATION_RULES vr ON ve.rule_id = vr.rule_id
               LEFT JOIN CHALLANS c ON ve.event_id = c.event_id
               WHERE ve.plate_no = %s
               ORDER BY ve.event_timestamp DESC""",
            (plate_no,)
        )
        
        violations = cursor.fetchall()
        
        # Convert date/datetime objects to strings
        for violation in violations:
            if violation.get('event_timestamp'):
                violation['event_timestamp'] = violation['event_timestamp'].isoformat()
            if violation.get('issue_date'):
                violation['issue_date'] = violation['issue_date'].isoformat()
            if violation.get('due_date'):
                violation['due_date'] = violation['due_date'].isoformat()
            if violation.get('paid_at'):
                violation['paid_at'] = violation['paid_at'].isoformat()
            # Convert Decimal to float
            if violation.get('base_fine_amount'):
                violation['base_fine_amount'] = float(violation['base_fine_amount'])
            if violation.get('total_amount'):
                violation['total_amount'] = float(violation['total_amount'])
        
        # Calculate summary statistics
        total_violations = len(violations)
        unpaid_challans = sum(1 for v in violations if v.get('payment_status') == 'Unpaid')
        total_unpaid_amount = sum(
            float(v['total_amount']) for v in violations 
            if v.get('payment_status') == 'Unpaid' and v.get('total_amount')
        )
        
        return {
            "message": "Vehicle search successful",
            "vehicle": vehicle,
            "summary": {
                "total_violations": total_violations,
                "unpaid_challans": unpaid_challans,
                "total_unpaid_amount": total_unpaid_amount
            },
            "violations": violations
        }
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )
    finally:
        if cursor:
            cursor.close()
        if conn and conn.open:
            conn.close()

File: backend/routes/vehicles.js
const express = require('express');
const db = require('../db');
const router = express.Router();

// GET /api/vehicles/search/:plateNo
router.get('/search/:plateNo', async (req, res) => {
  const plate = req.params.plateNo.toUpperCase().replace(/\s+/g, '');
  try {
    const [[vehicle]] = await db.execute(
      `SELECT v.plate_no, v.vehicle_type, v.vehicle_model, v.owner_name,
              v.owner_type, v.registered_at, v.citizen_id,
              c.full_name AS citizen_name, c.email AS citizen_email,
              c.trust_score
       FROM VEHICLES v
       LEFT JOIN CITIZENS c ON v.citizen_id = c.citizen_id
       WHERE REPLACE(v.plate_no,' ','') = ?`, [plate]
    ).catch(() => [[null]]);

    if (!vehicle) {
      return res.status(404).json({ detail: 'Vehicle not found in database.' });
    }

    const [violations] = await db.execute(
      `SELECT r.report_id,
              r.date_reported AS event_timestamp,
              r.location_address, r.status, r.description AS violation_description,
              ch.challan_id, ch.total_amount, ch.payment_status,
              ch.issue_date, ch.due_date,
              vr.rule_name, vr.rule_code, vr.severity, vr.base_fine_amount
       FROM REPORTS r
       LEFT JOIN VIOLATION_EVENTS ve ON r.report_id = ve.report_id
       LEFT JOIN CHALLANS ch ON ve.event_id = ch.event_id
       LEFT JOIN VIOLATION_RULES vr ON ve.rule_id = vr.rule_id
       WHERE REPLACE(r.plate_no,' ','') = ?
       ORDER BY r.date_reported DESC`, [plate]
    );

    const unpaid = violations.filter(v => ['Unpaid','Overdue'].includes(v.payment_status));
    res.json({
      vehicle,
      violations,
      summary: {
        total_violations: violations.length,
        unpaid_challans: unpaid.length,
        total_unpaid_amount: unpaid.reduce((s, v) => s + parseFloat(v.total_amount || 0), 0)
      }
    });
  } catch (err) {
    console.error('Vehicle search error:', err);
    res.status(500).json({ detail: 'Search failed: ' + err.message });
  }
});

// POST /api/vehicles/register
router.post('/register', async (req, res) => {
  const { plate_no, vehicle_model, vehicle_type, owner_name, owner_type, citizen_id } = req.body;
  if (!plate_no) return res.status(400).json({ error: 'plate_no required.' });
  try {
    const [[existing]] = await db.execute(
      `SELECT plate_no FROM VEHICLES WHERE plate_no=?`, [plate_no.toUpperCase()]
    );
    if (existing) return res.status(409).json({ error: 'Vehicle already registered.', plate_no: existing.plate_no });
    await db.execute(
      `INSERT INTO VEHICLES (plate_no, vehicle_model, vehicle_type, owner_name, owner_type, citizen_id)
       VALUES (?,?,?,?,?,?)`,
      [plate_no.toUpperCase(), vehicle_model || 'Unknown', vehicle_type || 'Car',
       owner_name || 'Unknown', owner_type || 'Individual', citizen_id || null]
    );
    res.status(201).json({ message: 'Vehicle registered successfully.', plate_no: plate_no.toUpperCase() });
  } catch (err) {
    console.error('Vehicle register error:', err);
    res.status(500).json({ error: 'Registration failed: ' + err.message });
  }
});

// PUT /api/vehicles/:plateNo  — update vehicle details
router.put('/:plateNo', async (req, res) => {
  const plate = req.params.plateNo.toUpperCase();
  const { vehicle_model, vehicle_type, owner_name, owner_type, citizen_id } = req.body;
  try {
    const [result] = await db.execute(
      `UPDATE VEHICLES
       SET vehicle_model = COALESCE(?, vehicle_model),
           vehicle_type  = COALESCE(?, vehicle_type),
           owner_name    = COALESCE(?, owner_name),
           owner_type    = COALESCE(?, owner_type),
           citizen_id    = COALESCE(?, citizen_id)
       WHERE plate_no = ?`,
      [vehicle_model || null, vehicle_type || null, owner_name || null,
       owner_type || null, citizen_id || null, plate]
    );
    if (result.affectedRows === 0) return res.status(404).json({ error: 'Vehicle not found.' });
    res.json({ message: 'Vehicle updated successfully.', plate_no: plate });
  } catch (err) {
    console.error('Vehicle update error:', err);
    res.status(500).json({ error: 'Update failed: ' + err.message });
  }
});

// DELETE /api/vehicles/:plateNo
router.delete('/:plateNo', async (req, res) => {
  const plate = req.params.plateNo.toUpperCase();
  try {
    const [result] = await db.execute(`DELETE FROM VEHICLES WHERE plate_no=?`, [plate]);
    if (result.affectedRows === 0) return res.status(404).json({ error: 'Vehicle not found.' });
    res.json({ message: 'Vehicle removed successfully.' });
  } catch (err) {
    console.error('Vehicle delete error:', err);
    res.status(500).json({ error: 'Delete failed: ' + err.message });
  }
});

// GET /api/vehicles/citizen/:citizenId  — all vehicles owned by a citizen
router.get('/citizen/:citizenId', async (req, res) => {
  try {
    const [rows] = await db.execute(
      `SELECT plate_no, vehicle_model, vehicle_type, owner_name, owner_type, registered_at
       FROM VEHICLES WHERE citizen_id = ? ORDER BY registered_at DESC`,
      [req.params.citizenId]
    );
    res.json({ vehicles: rows, count: rows.length });
  } catch (err) {
    console.error('Citizen vehicles error:', err);
    res.status(500).json({ error: 'Failed: ' + err.message });
  }
});

module.exports = router;
